mrmustard/physics/representations.py

```diff
@@ -482,20 +512,71 @@ def __getitem__(self, idx: int | tuple[int, ...]) -> Fock:
         new._contract_idxs = idx
         return new
 
-    def __matmul__(self, other: Fock) -> Fock:
+    def __matmul__(self, other: Union[Bargmann, Fock]) -> Fock:
         r"""
         Implements the inner product of ansatze across the marked indices.
 
-        Batch:
-        The new Fock holds the tensor product batch of them.
+        If ``other`` is ``Fock``, the two representations are automatically reduced
+        before being contracted. The array of the returned representation has the largest
+        possible dimension along every axis.
+
+        If ``other`` is ``Bargmann``, it is converted to ``Fock`` before the contraction.
+
+        Args:
+            other: Another representation.
 
-        Order of index:
-        The new Fock's order is arranged as uncontracted elements in self and then other.
+        Returns:
+            A ``Fock``representation.
         """
-        axes = [list(self._contract_idxs), list(other._contract_idxs)]
+        idx_s = list(self._contract_idxs)
+        idx_o = list(other._contract_idxs)
+
+        # if ``other`` is ``Bargmann``, convert it to ``Fock``
+        if isinstance(other, Bargmann):
+            from .converters import to_fock  # pylint: disable=import-outside-toplevel
+
+            # set same shape along the contracted axes, and default shape along the
+            # axes that are not being contracted
+            shape = [settings.AUTOCUTOFF_MAX_CUTOFF for _ in range(len(other.b[0]))]
```

Comment:
Reminder for self: implement `__len__` in Wires, so we can do `range(len(other.wires))`